packages/core/event.js

```diff
@@ -126,33 +127,130 @@ const stringOrFallback = (str, fallback) => typeof str === 'string' && str ? str
+BugsnagEvent.getStacktrace = function (error, errorFramesToSkip, backtraceFramesToSkip) {
   if (hasStack(error)) return ErrorStackParser.parse(error).slice(errorFramesToSkip)
-  // in IE11 a new Error() doesn't have a stacktrace until you throw it, so try that here
+  // error wasn't provided or didn't have a stacktrace so try to walk the callstack
   try {
-    throw error
+    return filter(StackGenerator.backtrace(), frame =>
+      (frame.functionName || '').indexOf('StackGenerator$$') === -1
+    ).slice(1 + backtraceFramesToSkip)
   } catch (e) {
-    if (hasStack(e)) return ErrorStackParser.parse(error).slice(1 + generatedFramesToSkip)
-    // error wasn't provided or didn't have a stacktrace so try to walk the callstack
-    try {
-      return filter(StackGenerator.backtrace(), frame =>
-        (frame.functionName || '').indexOf('StackGenerator$$') === -1
-      ).slice(1 + generatedFramesToSkip)
-    } catch (e) {
-      return []
-    }
+    return []
   }
 }
 
-BugsnagEvent.ensureEvent = function (eventOrError, errorFramesToSkip = 0, generatedFramesToSkip = 0) {
-  // notify() can be called with an Event object. In this case no action is required
-  if (eventOrError.__isBugsnagEvent) return eventOrError
+BugsnagEvent.create = function (maybeError, tolerateNonErrors, handledState, component, errorFramesToSkip = 0, logger) {
+  const [error, internalFrames] = normaliseError(maybeError, tolerateNonErrors, component, logger)
+  let event
   try {
-    const stacktrace = BugsnagEvent.getStacktrace(eventOrError, errorFramesToSkip, 1 + generatedFramesToSkip)
-    return new BugsnagEvent(eventOrError.name, eventOrError.message, stacktrace, undefined, eventOrError)
+    const stacktrace = BugsnagEvent.getStacktrace(
+      error,
+      // if an error was created/throw in the normaliseError() function, we need to
+      // tell the getStacktrace() function to skip the number of frames we know will
+      // be from our own functions. This is added to the number of frames deep we
+      // were told about
+      internalFrames > 0 ? 1 + internalFrames + errorFramesToSkip : 0,
+      // if there's no stacktrace, the callstack may be walked to generated one.
+      // this is how many frames should be removed because they come from our library
+      1 + errorFramesToSkip
+    )
+    event = new BugsnagEvent(error.name, error.message, stacktrace, handledState, maybeError)
   } catch (e) {
-    return new BugsnagEvent(eventOrError.name, eventOrError.message, [], undefined, eventOrError)
+    event = new BugsnagEvent(error.name, error.message, [], handledState, maybeError)
+  }
+  if (error.name === 'InvalidError') {
+    event.addMetadata(`${component}`, 'non-error parameter', makeSerialisable(maybeError))
+  }
+  return event
+}
+
+const makeSerialisable = (err) => {
+  if (err === null) {
+    return 'null'
+  } else if (err === undefined) {
+    return 'undefined'
+  } else if (isError(err)) {
```

Comment:
Struggling to understand this part. Under what conditions do we have both `error.name === 'InvalidError'` and `isError(err) === true`?

Reply:
I need to take another look at this. This logic used to live in the rejection handler and it was for when the rejection reason did not have a `stack`, but it was an error – such as a `DOMException`.

Reply:
I moved some of this logic back into the promise rejection handler plugin, as it was only ever used by that.

Reply:
To answer your original question:

> Under what conditions do we have both `error.name === 'InvalidError'` and `isError(err) === true`?

I think it can never have been the case.